src/ui/utils/tokenManager.tsx

```diff
@@ -0,0 +1,150 @@
+import { AppState } from "@auth0/auth0-react/dist/auth0-provider";
+import jwt_decode from "jwt-decode";
+import React, { ReactNode } from "react";
+import { useHistory } from "react-router-dom";
+import { assert, defer, Deferred } from "protocol/utils";
+
+const domain = "webreplay.us.auth0.com";
+const clientId = "4FvFnJJW4XlnUyrXQF8zOLw6vNAH1MAo";
+const audience = "hasura-api";
+const tokenRefreshSecondsBeforeExpiry = 60;
+
+const {
+  location: { origin, pathname },
+} = window;
+
+export interface TokenState {
+  token?: string;
+  error?: any;
+}
+
+type TokenListener = (state: TokenState) => void;
+
+class TokenManager {
+  private auth0Client: Auth0ContextInterface | undefined;
+  private deferredState = defer<TokenState>();
+  private isTokenRequested = false;
+  private refreshTimeout: number | undefined;
+  private listeners: TokenListener[] = [];
+
+  Auth0Provider = ({ children }: { children: ReactNode }) => {
+    const history = useHistory();
+
+    const onRedirectCallback = (appState: AppState) => {
+      history.push(appState?.returnTo || window.location.pathname);
+    };
+
+    return (
+      <Auth0Provider
+        domain={domain}
+        clientId={clientId}
+        audience={audience}
+        redirectUri={origin + pathname}
+        onRedirectCallback={onRedirectCallback}
+        cacheLocation="localstorage"
+        prompt="select_account"
+        useRefreshTokens={true}
+      >
+        <Auth0Context.Consumer>
+          {auth0Client => {
+            this.auth0Client = auth0Client;
+            setTimeout(() => this.update(false), 0);
+            return null;
+          }}
+        </Auth0Context.Consumer>
+        {children}
+      </Auth0Provider>
+    );
+  };
+
+  addListener(listener: TokenListener) {
+    this.listeners.push(listener);
+  }
+
+  removeListener(listener: TokenListener) {
+    const index = this.listeners.indexOf(listener);
+    if (index >= 0) {
+      this.listeners.splice(index, 1);
+    }
+  }
+
+  getToken() {
+    return this.deferredState.promise;
+  }
+
+  /**
+   * This method must be called before using Auth0's loginWithPopup() method.
+   */
+  reset() {
+    this.deferredState = defer<TokenState>();
+    this.isTokenRequested = false;
+    if (this.refreshTimeout !== undefined) {
+      clearTimeout(this.refreshTimeout);
+      this.refreshTimeout = undefined;
+    }
+  }
+
+  private async update(refresh: boolean) {
+    if (!this.auth0Client || this.auth0Client.isLoading) {
+      return;
+    }
+
+    if (this.auth0Client.isAuthenticated) {
+      if (this.isTokenRequested) {
+        if (refresh) {
+          this.reset();
+        } else {
+          return;
+        }
+      }
+
+      this.isTokenRequested = true;
+      const deferredState = this.deferredState;
+      try {
+        const token = await this.fetchToken(refresh);
+
+        this.setState({ token }, deferredState);
+        if (deferredState === this.deferredState) {
```

Comment:
Is there concern that this will change as a side-effect of triggering the listeners in `setState`, or only that it will change during fetching? If it's just fetching, it might be clearer to have
```
const deferredState = this.deferredState;
try {
  const token = await this.fetchToken(refresh);
  if (deferredState === this.deferredState) {
    return;
  }
  this.setState({ token });
  this.setupTokenRefresh(token):
} catch (e) {
  this.setState({ error: e });
}
assert(deferredState === this.deferredState);
```
to remove the need for the `deferredState` param.

Reply:
I was only concerned about it changing during fetching. But in that case I don't want to simply return because that would mean a promise returned by `getToken()` in the meantime would never be settled. Also, your code could still end up trying to resolve a Deferred twice if `this.deferredState` changed while fetching and the fetching then throws.